### ISSUE TITLE
Nuxt development server continiously restarts with `--https`

### ISSUE THREAD

**Issue by nandi95** [nuxt/framework#1796]
### Environment

```
------------------------------
- Operating System: `Darwin`
- Node Version:     `v16.11.1`
- Nuxt Version:     `3.0.0-27274229.29599f0`
- Package Manager:  `npm@8.1.3`
- Bundler:          `Vite`
- User Config:      `publicRuntimeConfig`, `ssr`, `target`, `meta`, `build`, `css`, `buildModules`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------
```

### Describe the bug

the development server seems to restart all the time

### Reproduction

steps:
- install fresh project with `nuxi init`
- generate ssl key and certificate
- updated dev script to `nuxi dev --host domain.test --https --ssl-key key.pem --ssl-cert cert.pem`

### Additional context

A similar issue has been described here: https://github.com/nuxt/framework/issues/1697
recreating the pkg lock and/or adding the `std-env` package does not help

**Comment by pi0** [comment]
Additional context: Browser logs:

> client.ts:28 WebSocket connection to 'wss://localhost:24678/_nuxt/' failed: 

It seems a vite issue. You can use `vite: false` in `nuxt.config` as a workaround

**Comment by nandi95** [comment]
Hmm, I'm not getting any logs in the browser nor the terminal

**Comment by pi0** [comment]
> Hmm, I'm not getting any logs in the browser nor the terminal

They go away quickly you need to check "preserve logs" to see them 🙈

**Comment by pi0** [comment]
@nandi95 Can you please try this to see if works:

```js
export default defineNuxtConfig({
  vite: {
    server: {
      hmr: {
        protocol: 'ws'
      }
    }
  }
})
```

**Comment by dbhagen** [comment]
Having a similar issue behind a Caddy reverse proxy.
Turning off Vite works:
```bash
npx nuxi info                   
Nuxt CLI v3.0.0-27274229.29599f0                                                                                                                                                 09:55:49
RootDir: /Users/danielhagen/Projects/cashstudio-nuxt3                                                                                                                            09:55:50
Nuxt project info:                                                                                                                                                               09:55:50

------------------------------
- Operating System: `Darwin`
- Node Version:     `v14.18.1`
- Nuxt Version:     `3.0.0-27274229.29599f0`
- Package Manager:  `yarn@1.22.17`
- Bundler:          `Webpack`
- User Config:      `vite`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------
```
Console log
```
Navigated to https://localhost.test.domain/
runtime-core.esm-bundler.js?5c40:893 <Suspense> is an experimental feature and its API will likely change.
client.js?aa7f:95 [HMR] connected
```
With Vite in WSS mode:
```bash
npx nuxi info                                 
Nuxt CLI v3.0.0-27274229.29599f0                                                                                                                                                 09:59:55
RootDir: /Users/danielhagen/Projects/cashstudio-nuxt3                                                                                                                            09:59:55
Nuxt project info:                                                                                                                                                               09:59:55

------------------------------
- Operating System: `Darwin`
- Node Version:     `v14.18.1`
- Nuxt Version:     `3.0.0-27274229.29599f0`
- Package Manager:  `yarn@1.22.17`
- Bundler:          `Vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------
```
Console log
```
Navigated to https://localhost.test.domain/
client.ts:22 [vite] connecting...
runtime-core.esm-bundler.js:893 <Suspense> is an experimental feature and its API will likely change.
client.ts:28 WebSocket connection to 'wss://localhost.test.domain:24678/_nuxt/' failed: 
(anonymous) @ client:188
client.ts:224 [vite] server connection lost. polling for restart...
```

Running Vite in a non-secure WebSocket, as mentioned above by @pi0, results in an error.
```
Navigated to https://localhost.test.domain/
client.ts:22 [vite] connecting...
client.ts:28 Mixed Content: The page at 'https://localhost.test.domain/' was loaded over HTTPS, but attempted to connect to the insecure WebSocket endpoint 'ws://localhost.test.domain:24678/_nuxt/'. This request has been blocked; this endpoint must be available over WSS.
(anonymous) @ client.ts:28
2localhost.test.domain/:1 Uncaught DOMException: Failed to construct 'WebSocket': An insecure WebSocket connection may not be initiated from a page loaded over HTTPS.
    at https://localhost.test.domain/_nuxt/@vite/client:188:16
```

**Comment by nandi95** [comment]
that doesn't seem to work when using custom configuration like
`nuxi dev --host website.test --https --ssl-key key.pem --ssl-cert cert.pem --port 3000`
Tried with updating the `server` key with the values from the command arguments to no avail

**Comment by misaon** [comment]
I have a Docker stack and an Nginx reverse proxy in it, which proxy pass "/" to "http://client:3000" when i visit https://web.localhost domain.  So.... :

If I use Nuxt3 without Docker and reverse proxy, go to port http://localhost:3000, nuxt.config `vite->server->hmr` is loaded properly. If I go to https://web.localhost `vite->server->hmr` **is ignored** 😢

Some idea?

**Comment by vicvillalobos** [comment]
> I have a **Docker** stack and an **Nginx reverse proxy** in it, which proxy pass "/" to "http://client:3000" when i visit https://web.localhost domain. So....
> 
> If I use Nuxt3 **without Docker and reverse proxy**, go to http://localhost:3000, nuxt.config `vite->server->hmr` is loaded properly 👍🏼. If I go to https://web.localhost `vite->server->hmr` options **are ignored**! 👎🏼 😢
> 
> Some idea?

In case anyone stumbles upon this issue, I solved it by exposing the 24678 port on docker.

**Comment by JamieMcDonnell** [comment]
I found a simple solution to this in my nuxt.config.ts - my vite settings now look like this:
```
vite: {
    server: {
      port: 443,
      https: true,
      hmr: {
        protocol: 'wss'
      }
    }
  }
 ```

It was protocol: 'wss' which did the trick - it needs to be secure, as it is running on https...

**Comment by ol1s** [comment]
> In case anyone stumbles upon this issue, I solved it by exposing the 24678 port on docker.

Any way of configuring custom ports - I believe you can on Vite but not sure if this translates onto nitro?

**Comment by Vejron** [comment]
When running **nuxi** in https you also need to tell **vite** where the certificates is. Otherwise you get the constant refreshing issue.
In nuxt config something like this should suffice
```
vite: {
    server: {
      https: {
        key: fs.readFileSync('cert/localhost+2-key.pem'),
        cert: fs.readFileSync('cert/localhost+2.pem'),
      },
      hmr: {
        protocol: 'wss'
      }
    }
  },
```

**Comment by JamieMcDonnell** [comment]
You might need to use "sudo" when starting the server too, at least I did.

**Comment by LaCocoRoco** [comment]
If anyone uses Caddy Revers Proxy and Https in an Development Environment, this fixed my problem.

**Config Nuxt**
```typescript
import { defineNuxtConfig } from 'nuxt';

export default defineNuxtConfig({
  vite: {
    server: {
      hmr: {
        protocol: 'wss',
        clientPort: 443,
      },
    },
  },
});
```

**Config Caddy**
```
dev.domain.com {
  @websockets {
    header Connection Upgrade
    header Upgrade websocket
  }
  
  reverse_proxy @websockets localhost:24678
  reverse_proxy localhost:3000
}
```

**Comment by shtse8** [comment]
[vite team fixed the issue.](https://github.com/vitejs/vite/pull/8650)

**Comment by nandi95** [comment]
Now just need to wait until vite reaches v3 and nuxt starts to use it.
Hopefully this websocket protocol infer will fix it but in the meantime, I figured that this error only occurs when I add the argument `--host domain.test`. Then it refreshes. HRM and everything works fine with all other arguments.
Also worth noting that deleted the `.nuxt` folder when trying which argument breaks it, as if I left it, it teased out this issue: https://github.com/nuxt/framework/issues/5183

**Comment by nandi95** [comment]
with RC-7 out (vite upgraded) there is still an issue. It just keep polling.
Firefox won't connect to wss on a different port. If I set the port to same as website, it restarts a couple times but websocket isn't in the network tab and hmr doesn't work.
if I set the protocol to wss manually it doesn't restart but it doesn't set up hmr either.
Reproduction: https://github.com/nandi95/nuxt-starter

**Comment by TheDutchCoder** [comment]
I'm dealing with a tangentially related issue, where if you use https and a custom host name (say `local.mydomain.com` mapped to 127.0.0.1 through your hosts file), then HRM fails on the `ws://local/mydomain.com` connection complaining about having to use `wss`. If you actually set `wss` as the protocol it will also not work.

As an aside, the typings on defineNuxtConfig for `vite` are incomplete and will throw errors when using properties like `server`:
<img width="744" alt="image" src="https://user-images.githubusercontent.com/1668207/190625903-d4e5c548-f322-4745-bc35-7a9479267247.png">

**Comment by danielroe** [comment]
(The `vite.server` options `host` and `port` should not be configured directly because they need to be set by Nuxt. You configure them with `--port` and `--host` command line arguments.)

**Comment by nandi95** [comment]
Even when configuring using the cli it's still an issue. See the reproduction.

**Comment by danielroe** [comment]
Not disputing that, just explaining why we don't want configuring `vite.server.host` directly :-)

**Comment by TheDutchCoder** [comment]
Okay, so this might either be a Vite internal issue, or the way vite config is used in Nuxt.
Setting something like
```js
    server: {
      hmr: {
        host: 'local.reinier.remx.xyz',
        protocol: 'wss',
      },
    },
```
Has no effect, it just says the websocket has failed with no specific message:
<img width="545" alt="image" src="https://user-images.githubusercontent.com/1668207/190635776-defe9f48-8afb-44c7-a9a7-96ce071a0ec8.png">

**Comment by nandi95** [comment]
After https://github.com/nuxt/framework/pull/7680 this now works. with a custom certificate and key the host and wss will now work over https and wss.

However, when using only `--https --host` without the key and certificate the site will not load:
<img width="592" alt="Screenshot 2022-09-21 at 09 28 31" src="https://user-images.githubusercontent.com/41805560/191455083-359c85bd-4fa2-4b2c-a76a-21970ddf9db9.png">

Is this worth reopening as a new issue?
Is this perhaps an issue within [unjs/listhen](https://github.com/unjs/listhen)?

**Comment by danielroe** [comment]
@nandi95 Good idea. I'll close this one as resolved, and yes, let's raise as a new issue.

**Comment by TheDutchCoder** [comment]
> After #7680 this now works. with a custom certificate and key the host and wss will now work over https and wss.
> 
> However, when using only `--https --host` without the key and certificate the site will not load: <img alt="Screenshot 2022-09-21 at 09 28 31" width="592" src="https://user-images.githubusercontent.com/41805560/191455083-359c85bd-4fa2-4b2c-a76a-21970ddf9db9.png">
> 
> Is this worth reopening as a new issue? Is this perhaps an issue within [unjs/listhen](https://github.com/unjs/listhen)?

I have the same issue and tried to resolve it with pi0 yesterday on discord.

It works for him, but I have a suspicion that he, at some point, added the certificates to his OS.

Chrome blocks unsigned certificates.

**Comment by pi0** [comment]
> added the certificates to his OS.

Each time you restart nuxi, there is a new root certificate chain generated. 

I will try to investigate on a couple of more browsers and operating systems.

**Comment by danielroe** [comment]
On Chrome you can override :

![CleanShot 2022-09-21 at 11 16 14](https://user-images.githubusercontent.com/28706372/191479313-fdcea757-7d69-4ba4-ba28-581632d03e42.png)

But there's no obvious way to override in FF.

**Comment by pi0** [comment]
@nandi95 Please try on RC.11. In previous RCs, HTTPS support had two main issues:
- HTTPS Certificate wasn't secure enough to be able bypass in chrome (FF seems needs other changes)
- HMR port serving ws:// could not be reached from secure localhost neither we were sharing same certificate (that is bypassed in chrome).

**Comment by Reivax47** [comment]
> 

This workaround works for me, but only with Chrome. Firefox can't connect and I get the message 'server connection lost. polling for restart'. I tried with the 'HelloWorld' Nuxt3 app. Caddy and Nuxt app are dockerized.

If I use a Vite app (without Nuxt3), Vite HMR works fine...

On the left : Chrome. On the right : Firefox.
![image](https://user-images.githubusercontent.com/47529668/204741002-b3e44247-627c-4a37-844c-f7bc2b03e5a7.png)

About my config :  

```
Nuxi 3.0.0                                                                                                                                                             08:07:24
                                                                                                                                                                       08:07:25
RootDir: /root/front                                                                                                                                                   08:07:25
Nuxt project info:                                                                                                                                                     08:07:25

------------------------------
- Operating System: `Linux`
- Node Version:     `v16.18.1`
- Nuxt Version:     `3.0.0`
- Nitro Version:    `1.0.0`
- Package Manager:  `npm@8.19.2`
- Builder:          `vite`
- User Config:      `vite`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

```
The Nuxt.config.ts 
                                                             
```
// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
vite:{
        server: {
        hmr:{
                        protocol: 'wss',
                        clientPort: 443,
        },
  },
},
})

```
The Caddyfile : 

```
                                                                            
komelia.fr {

  @websockets {
    header Connection Upgrade
    header Upgrade websocket
  }

        reverse_proxy @websockets front:24678
        reverse_proxy front:3000

}
```

NuxtApp docker file : 

```
#Front
FROM node:16.18-bullseye
RUN apt -y update
RUN apt -y upgrade
RUN apt-get -y install  curl vim


# Montage des documents partagés
COPY ./nuxt-app/ /root/front/
WORKDIR /root/front
RUN npm install

ENV NUXT_HOST=0.0.0.0
ENV NUXT_PORT=3000
CMD ["npm", "run", "dev"]


```
As it's works fine with Chrome, I'm guessing Caddyfile is allright... ?
I'm running the app with " npm run dev".

**Comment by tasiotas** [comment]
@Reivax47 
Try this config, it works well for me in Dockerizer enviro.

nuxt.config.ts
```ts
  vite: {
    server: {
      hmr: {
        protocol: 'wss',
        clientPort: 443,
        path: 'hmr/',
      },
    },
}
```

Caddyfile
```
localhost {
	tls /certs/localhost.crt /certs/localhost.key # up to you how you handle certs

	handle /_nuxt/hmr/ {
		reverse_proxy http://nuxt:24678 # nuxt is the name of my service in docker-compose.yml
	}

	handle {
		reverse_proxy http://nuxt:3000
	}

}

**Comment by Reivax47** [comment]
@tasiotas That did the tricks ! 
Thanks a lot.

**Comment by rrolla** [comment]
Thanks, @tasiotas with your configuration I manage to get it working also in nuxt3 with traefik

`nuxt.config.ts`
```ts
  vite: {
    server: {
      hmr: {
        protocol: 'wss',
        port: 22300,
        clientPort: 443,
        path: 'hmr/',
        timeout: 3,
      },
    },
  },
```

Docker configuration most important is this 
```yml
traefik.http.routers.vitesocket.rule=Host(`docker.dev.gradio.lv`) && PathPrefix(`/_nuxt/hmr/`)
```

And to expose two ports, otherwise traefik cannot create new service for each port


`docker-compose.yml`

```yml
  gradiohead:
    container_name: gradio_head
    hostname: v5.docker.dev.gradio.lv
    image: node:14-alpine
    labels:
      - traefik.enable=true
      - traefik.http.routers.gradiohead_http.service=gradiohead
      - traefik.http.routers.gradiohead_http.rule=Host(`docker.dev.gradio.lv`, `v5.docker.dev.gradio.lv`)
      - traefik.http.routers.gradiohead_http.entrypoints=http
      - traefik.http.routers.gradiohead_http.middlewares=redirect-to-https

      - traefik.http.middlewares.redirect-to-https.redirectscheme.scheme=https
      - traefik.http.routers.gradiohead.service=gradiohead
      - traefik.http.services.gradiohead.loadbalancer.server.port=3000
      - traefik.http.routers.gradiohead.rule=Host(`docker.dev.gradio.lv`, `v5.docker.dev.gradio.lv`)
      - traefik.http.routers.gradiohead.entrypoints=https
      - traefik.http.routers.gradiohead.tls=true

      - traefik.http.routers.vitesocket.service=vitesocket
      - traefik.http.services.vitesocket.loadbalancer.server.port=22300
      - traefik.http.routers.vitesocket.rule=Host(`docker.dev.gradio.lv`) && PathPrefix(`/_nuxt/hmr/`)
      - traefik.http.routers.vitesocket.entrypoints=http,https
      - traefik.http.routers.vitesocket.tls=true
    expose:
      - 3000
      - 22300
    command: >
        /bin/sh -c "yarn && yarn dev"
```

And you will get response status `101`

![image](https://user-images.githubusercontent.com/3932546/209674957-4765e9cf-4169-4508-ad54-bc92ee19333e.png)

**Comment by pi0** [comment]
PR welcome to improve docs for reverse-proxy (nginx, traefik) in development.
